src/marqo/core/structured_vespa_index/structured_add_document_handler.py

```
@@ -58,12 +69,25 @@ def _handle_field(self, marqo_doc, field_name, field_content):
         self._validate_field(field_name, field_content)
         content = self.tensor_fields_container.collect(
             marqo_doc[MARQO_DOC_ID], field_name, field_content,
-            self._infer_field_type
+            self.marqo_index.field_map[field_name].type
         )
         marqo_doc[field_name] = content
 
-    def _infer_field_type(self, field_name:str, field_content: Any) -> FieldType:
-        return self.marqo_index.field_map[field_name].type
+    def _infer_modality(self, tensor_field: TensorField) -> Modality:
+        if tensor_field.field_type == FieldType.Text:
+            return Modality.TEXT
+
+        url = tensor_field.field_content
+        try:
+            modality = infer_modality(url, self.add_docs_params.media_download_headers)
```

Could this lead to inconsistent behaviour? I think previously we just process the content based on the given modality, and rely on the preprocessor to raise an error.

---

That's correct, and we cannot do that now since Inference does not know about the field type. The good thing about doing it earlier is that if there's any mismatch, we fail early and do not send to Inference.